app/src/androidTest/java/com/github/campus_capture/bootcamp/FirebaseBackendTest.java

```diff
@@ -0,0 +1,262 @@
+import org.junit.Test;
+import org.junit.runner.RunWith;
+
+import java.security.spec.ECField;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.TimeUnit;
+
+@RunWith(AndroidJUnit4.class)
+public class FirebaseBackendTest {
+
+    static FirebaseDatabase database = null;
+
+    @BeforeClass
+    public static void init_firebase_emulator() {
+
+        try{
+            AppContext context = AppContext.getAppContext();
+            database = context.getFirebaseDB();
+            Log.d("FirebaseBackendTest", "enable emulator");
+            database.useEmulator("10.0.2.2", 9000);
+
+        } catch (Exception e) {
+            Log.e("FirebaseBackendTest", e.toString());
+        }
+    }
+
+    @Before
+    public void clear_firebase_database(){ database.getReference().setValue(null);}
+
+    @Test
+    public void testVoteZone() {
+        // set database content
+        database.getReference().child("Users").child("testUserId").child("has_voted").setValue(false);
+        database.getReference().child("Zones").child("BC").child("IN").setValue(4);
+
+        BackendInterface b = new FirebaseBackend();
+
+        try {
+            Boolean result = b.voteZone("testUserId", Section.IN, "BC").get();
+            assertTrue(result);
+        }catch(Exception e){
+            Log.e("Error in test", e.toString());
+            assertTrue(false);
+        }
+
+        // check database content
+        CompletableFuture<Boolean> futureResultHasVoted = new CompletableFuture<>();
+        database.getReference().child("Users").child("testUserId").child("has_voted").get().addOnCompleteListener(new OnCompleteListener<DataSnapshot>() {
+            @Override
+            public void onComplete(@NonNull Task<DataSnapshot> task) {
+                if (!task.isSuccessful()) {
+                    fail();
+                }
+                else {
+                    futureResultHasVoted.complete( (Boolean)task.getResult().getValue() );
+                }
+            }
+        });
+
+        CompletableFuture<Long> futureResultVoteCount = new CompletableFuture<>();
+        database.getReference().child("Zones").child("BC").child("IN").get().addOnCompleteListener(new OnCompleteListener<DataSnapshot>() {
+            @Override
+            public void onComplete(@NonNull Task<DataSnapshot> task) {
+                if (!task.isSuccessful()) {
+                    fail();
+                }
+                else {
+                    futureResultVoteCount.complete( (Long)task.getResult().getValue() );
+                }
+            }
+        });
+
+        try{
+            assertTrue(futureResultHasVoted.get());
+            assertEquals(5, futureResultVoteCount.get().longValue());
+        }catch (Exception e){
+            fail();
+        }
+    }
+
+    @Test
+    public void testVoteZoneImpossibleAlreadyVoted() {
```

Comment:
Good thinking! I haven't even thought about locking off votes for players which already voted... since I was UI-focused I figured just disabling the button in the UI is enough but this is a lot more robust.